### ISSUE TITLE
False Positive error: "yield from" can't be applied to "Foo"  [misc] where Foo has explicit __iter__(self)

### ISSUE THREAD

**Issue by couling** [python/mypy#17305]
**Bug Report**

MyPy incorrectly reports an error when attempting to iterate over an object with explicit `__iter__(self)` method and no `__next__(self)`.  This is most easily hit using a generator.

**To Reproduce**

```python
from typing import Iterable

class Foo:
    def __iter__(self) -> Iterable[str]:
        yield "hello"
        yield "world"

def iter_foo() -> Iterable[str]:
    yield from Foo()

def print_foo() -> None:
    for word in Foo():
        print(word)
```

**Expected Behavior**

The above code should be fine

**Actual Behavior**

Two errors:
```
bar.foo.py:9: error: "yield from" can't be applied to "Foo"  [misc]
bar.foo.py:12: error: "Iterable[str]" has no attribute "__next__"  [attr-defined]
```
**Your Environment**

- Mypy version used: 1.10.0
- Mypy command-line flags: 
- Mypy configuration options from `mypy.ini` (and other config files):
  ``` 
   [tool.mypy]
   check_untyped_defs = true
   explicit_package_bases = true
   packages = [
       "bar",
   ]
   mypy_path = "source"
   ```
- Python version used: Python 3.11.4

**Comment by couling** [comment]
Sorry stupid type hint error. saw it as soon as I clicked subit.
